zerver/decorator.py

```diff
@@ -884,6 +887,30 @@ def rate_limit_user(request: HttpRequest, user: UserProfile, domain: str) -> Non
     RateLimitedUser(user, domain=domain).rate_limit_request(request)
 
 
+@cache_with_key(lambda: "tor_ip_addresses:", timeout=60 * 60)
+@circuit(failure_threshold=2, recovery_timeout=60 * 10)
```

Comment:
LGTM, just wondering if we should add a comment elaborating on why using `circuit` is needed? I don't fully understand how much of a benefit it brings and what the risks of not having it would be

Reply:
https://martinfowler.com/bliki/CircuitBreaker.html has more details about this pattern in general.

It's mild overkill here.  The idea is that we're doing a thing which is potentially time-intensive, and that we don't want to cache the output of when it fails, because, say, at startup, we may not have run the cron job yet.  Rather than continue to do hit disk every request for an hour, until the cron job runs, we can decide after a couple failed requests in a rowthat the situation is unlikely to change and stop looking for a bit.

It's overkill in this situation because the thing that calls isn't a network request with a multiple-second timeout, but rather going to local disk, which is milliseconds, not seconds.  It makes much more sense when we want to stop loading some remote system when it's likely-broken.  Here, it's effectively being used as a shorter local cache for failed disk checks.

Reply:
Makes sense, thanks for explaining!